### DISCUSSION TITLE
A few ideas from a world builder's perspective

### DISCUSSION THREAD

**Discussion by Irthwynd** [DavidRieman/WheelMUD#90] [Ideas]
I wasn't sure whether to post this. By the end it started to read like a wish list. It isn't. My inner world builder is just activated after reading the OLC story. :) This is a glimpse of how *I* would want to build a world. Whether these ideas would be universally beneficial for the core design of your engine, I'm not sure. Again, your OLC already looks great.

---
Creating an object.

I would want the option to add randomized adjectives to augment not only an item's name and description, but also the additional option for those adjectives to augment its properties or value. Reality is a bit random, right? I like that for us and I like that for a game. For example, suppose I want to create a ruby that belongs to a global loot table. I would set this item "ruby" to have a max "quality" of 10,000 (in this case the quality might directly correspond to its value.  Oh and these terms and numbers are entirely arbitrary). It would have two (or more) sets of randomized adjectives augmenting the ruby, and in a particular order. A (tiny, small, medium-sized, large, very large, enormous) (cracked, flawed, dull, translucent, clear, sparkling, luminous, brilliant, flawless) ruby. Tiny through enormous would then be set to represent a percentage of its value, with tiny representing, say, 1-10% of its max quality and enormous 95-100% of its max quality. Likewise, cracked representing 1-10% of max quality and flawless representing 98-100% max quality. Average the two and you get the quality and therefore the value of the ruby. The final value of a "tiny flawless ruby" would be roughly 50% of its max quality. Perhaps there could be the option to have one set of adjectives more heavily weighted in the equation. I would also set the ruby to have a greater chance to spawn with higher tier adjectives for higher level foes, and vice versa for lower level foes (or not spawn at all if they are too low). This would then set up situations for some items or materials to be refined, whether through skill or magic, to be used in crafting, magical rituals, etc. It might be part of a foundation of an economy of services among players.

Also to note, adjectives do not need to be random (which probably means you could just slap them in the item name), nor would they *have* to augment the value, properties, or attributes of an item.

This concept could be applied to weapons, armor, magical items, etc; however, this would not only affect its value, it could also augment the effectiveness of the item within given parameters. 

This concept could be applied to mobs as well, where a single mob within an area might be to some degree more or less effective at what it does. A short frail goblin might be slightly less dangerous than a short battle-scarred goblin.

---

For rooms, I would want to show both the region and room names. I would want the option to clearly display the time of day, the season, the weather, and any other obvious celestial activity within a room or region (if all or any is applicable). I would want the option for the visible room description to reflect both the time of day and season (if applicable). I would want the option for every noun within a room description to be a virtual item, which could have its own description and builder-defined list of interactive verbs. I would like the option for builder-defined atmospheric "events" to trigger on some timer within a room or region. here is a simple example (I'll just write in my own format):

[King's Wood, deep within the wood on a beaten path]
The sun is at its zenith, it is autumn, there is light cloud cover, a waning gibbous moon is faintly visible on the western horizon.
Warm shafts of sunlight dance through a shifting russet and crimson-tipped canopy overhead.  A pale patchwork of birch stretches to the north and west from here, interrupted by a series of small moss-covered boulders not far from the path.  The way leads deeper into the southern wood overgrown with verdant groundcover.  
Obvious exits: south and northeast.

>look canopy
Slender birch rise overhead, creating a vibrant canopy of autumn hues.

>look boulders
What appeared at first to be a nondescript series of small boulders seems to have some semblance of order. You count twelve small boulders arranged in a rough circle. Aged and weather-worn, moss has nearly covered what appears to be strange grooves on each surface.

>look grooves (observation check successful)
Although worn and partially obscured by moss, you discern the grooves to be weathered and worn ancient runes.  Each bolder is carved differently, some angular and other curvilinear.

>look runes (lore/observation check unsuccessful)
You can make out the outline of the runes, but their purpose and meaning are unclear to you.

>The lilt of water flowing over rock can be heard in the distance. (atmospheric timed message)

**Comment by DavidRieman** [comment]
On part 1 (random stuff)...  Yes!  Random systems (including varying mobs and loot) can add _so much_ to a game.  It can totally make the difference between replaying that zone again being _willingly fun_ versus _aren't there more areas to play to get off this repetitive grind_?  I think we can all learn from some of the more modern roguelikes and hack-n-slashers that excel with replayability.

There's something just enjoyable about that phase of going through your recent spoils and comparing... "this one adds +20 HP and +1 dex, but this one is +2% resist to fire... hmm, which fits my character best?" and when you can feel confidence in smashing through that level _again_ and now have that periodic reward for each tiny upgrade to what you've already got...  Definitely got some psychology behind that periodic reward schedule that makes games addictive.

Except _this_ time you run into an elite battle-scarred hulking goblin surrounded by a magical cloud of living acid and brandishing a lightning sword and you got to run away...  To reconsider what gear you're bringing to the table to counter that, or call for a group to clear the threat with the chance of that sweet elite loot?  That's starting to get into some cool emergent gameplay territory.

As a text-based game engine, we have the advantage that we don't even have to represent such systems in a _graphically_ compelling and obvious way.  Despite that, IMO what holds MUDs back from doing this stuff is _still_ the technical complexity.  These ancient code bases built on harder languages just aren't as flexible and easy to change as I hope C# WheelMUD to be.  I do hope even our reference implementation(s) of a game will demonstrate some emergent gameplay aspects along these lines.

**Comment by DavidRieman** [comment]
On part 2 (room stuff)...

IIRC our prior forums had some discussion about the role of weather and game time and such in a game.  We've got weather systems on the roadmap.  I don't recall much (and I'm still making progress of weeding out fluff posts and finding what archived content I want to bring here to github, even in resurrected Discussions form).  But I do recall, whether from there or seeing some of the code, that there were ideas to have flags for rooms to determine whether you could see the effects of weather changes (like if a room is "outdoors" you might be getting rained on) and a game system could use that to gate whether a Druid or Shaman could conjure lightning, or whatever.  I hadn't thought much about it being presented as part of the room description, and possibly affecting the room description (such as the "sunlight dancing through" passage).  Some interesting challenges there.  There may be some ways the framework can help facilitate such things, that will take some more thought later.

Typically when I've seen weather and time systems in a MUD it has been periodic update texts that come in when the weather changes, or commands you could issue to ask what the weather/time in game currently is.  But I have never really been compelled to care much about these systems in the games I played; they were mostly just descriptive icing.  Graphical or otherwise.  Neither though would I have wanted to be at the whim of whether a room was outdoors or stormy or whatnot to determine whether my character could be effective though, so I wouldn't usually opt to play such a class, personally.  On the other hand, I do see room for someone to design a MUD world where I _would_ have cared.  Perhaps in my case, I would have been interested if certain mobs only came out in or after the rain (imagine worm-like creatures) or flying mobs avoiding the storms, or nocturnal creatures only to spawn at night...

We've got the starting point for a perception system that can factor in perception strength on multiple factors (vision, hearing, sonar, and so on) influencing whether you see what a thing _is_ versus "someone" and so on, and I can see daylight / weather influencing base perception values.  I'm not fully satisfied with where the system is at, but it's something to experiment and iterate on.

For what you describe at the end - yeah that is definitely a feature I'm strongly in favor of.  If you check out the [World Interaction User Story](http://htmlpreview.github.io/?https://github.com/DavidRieman/WheelMUD/blob/master/docs/Design/UserStory_WorldInteraction.htm), among other technical features it goes over, it covers what I've tentatively called "Furnishings".  Maybe "Adornments" or something would be a good word choice too.  A key feature would be that it would highlight the relevant keyword where it appears in the room description automatically, so there's no guessing about what you can examine.  Also, now that I think of it, we could _also_ mark up those texts as clickable links for clients who support such MXP features (discovered through Telnet negotiation) to automatically send that look command when the link is clicked!

**Comment by Irthwynd** [comment]
Oof. I wrote a long reply and it disappeared when I clicked comment. Bummer. I'll summarize.

I definitely had roguelike in mind when writing out my suggestions, and I am glad you generally agree! My goal as a world builder is to, well, create a world. I think emergent gameplay is necessary to achieve that. I've helped build a handful of MUDs, but only one had randomized elements. Not enough to meaningfully impact the gameplay, and eventually the game felt static and stale. I'd like to create a living and breathing world, which is why I also want room descriptions to be dynamic. I do not expect any codebase to have that functionality baked in. I'd have to code it myself, and I hope one day I am at the point where I can.

I have very fond memories from my first MUD (Gemstone IV). The world came to life after I wandered around reading the room descriptions. I came to see it as another world and my character was living within it. It truly was magical. Roleplay became a huge part of my life. I want to recreate that magic for others of a like mind. It'd take a lot of work, but a worthwhile effort for the right audience.

As for the "furnishings", awesome! Gemstone IV also uses links and I'd love to play around with that, too. It looks like what you're building is jivin' with me. I look forward to seeing where you take your project, and I appreciate that you're open to these kinds of discussions. :)
